File: bindings/clojure/src/proven/safe_string.clj
;; SPDX-License-Identifier: AGPL-3.0-or-later
;; SPDX-FileCopyrightText: 2025 Hyperpolymath

(ns proven.safe-string
  "Safe string operations with XSS and injection prevention."
  (:require [clojure.string :as str])
  (:import [java.net URLEncoder URLDecoder]
           [java.util Base64]
           [java.nio.charset StandardCharsets]))

(def ^:private html-entities
  {\& "&amp;"
   \< "&lt;"
   \> "&gt;"
   \" "&quot;"
   \' "&#x27;"})

(defn escape-html
  "Escape string for safe HTML output (XSS prevention)."
  [input]
  (apply str (map #(get html-entities % (str %)) input)))

(defn escape-sql
  "Escape string for SQL single quotes.
   Note: PREFER PARAMETERIZED QUERIES! This is for edge cases only."
  [input]
  (str/replace input "'" "''"))

(defn escape-js
  "Escape string for JavaScript string context."
  [input]
  (apply str
         (map (fn [c]
                (case c
                  \\ "\\\\"
                  \' "\\'"
                  \" "\\\""
                  \newline "\\n"
                  \return "\\r"
                  \tab "\\t"
                  \< "\\x3C"
                  \> "\\x3E"
                  (str c)))
              input)))

(defn escape-shell
  "Escape string for shell command (single-quote wrapping)."
  [input]
  (str "'" (str/replace input "'" "'\\''") "'"))

(defn escape-regex
  "Escape special characters for use in regex patterns."
  [input]
  (java.util.regex.Pattern/quote input))

(defn url-encode
  "URL-encode a string."
  [input]
  (URLEncoder/encode input (.name StandardCharsets/UTF_8)))

(defn url-decode
  "URL-decode a string. Returns nil on error."
  [input]
  (try
    (URLDecoder/decode input (.name StandardCharsets/UTF_8))
    (catch Exception _ nil)))

(defn base64-encode
  "Base64-encode a string."
  [input]
  (.encodeToString (Base64/getEncoder) (.getBytes input StandardCharsets/UTF_8)))

(defn base64-decode
  "Base64-decode a string. Returns nil on error."
  [input]
  (try
    (String. (.decode (Base64/getDecoder) input) StandardCharsets/UTF_8)
    (catch Exception _ nil)))

(defn truncate
  "Safely truncate a string to a maximum length."
  ([input max-length] (truncate input max-length "..."))
  ([input max-length suffix]
   (cond
     (<= max-length 0) ""
     (<= (count input) max-length) input
     (<= max-length (count suffix)) (subs input 0 max-length)
     :else (str (subs input 0 (- max-length (count suffix))) suffix))))

(defn strip-html
  "Strip HTML tags from string (basic - not a full parser)."
  [input]
  (str/replace input #"<[^>]*>" ""))

(defn alphanumeric?
  "Check if string contains only alphanumeric characters."
  [input]
  (and (not (str/blank? input))
       (re-matches #"^[a-zA-Z0-9]+$" input)))

(defn ascii?
  "Check if string contains only ASCII characters."
  [input]
  (every? #(<= 0 (int %) 127) input))

(defn sanitize
  "Sanitize string to contain only specified allowed characters."
  ([input] (sanitize input "a-zA-Z0-9_-"))
  ([input allowed]
   (let [pattern (re-pattern (str "[^" allowed "]"))]
     (str/replace input pattern ""))))

(defn remove-control-chars
  "Remove control characters from string."
  [input]
  (str/replace input #"[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]" ""))

(defn normalize-whitespace
  "Normalize whitespace (collapse multiple spaces, trim)."
  [input]
  (-> input
      (str/replace #"\s+" " ")
      str/trim))

(defn contains-suspicious-patterns?
  "Check if string looks like it contains injection attempts."
  [input]
  (let [lower-input (str/lower-case input)]
    (or (str/includes? lower-input "<script")
        (str/includes? lower-input "javascript:")
        (some? (re-find #"on\w+\s*=" lower-input))
        (some? (re-find #"union\s+select" lower-input))
        (some? (re-find #";\s*drop\s+table" lower-input)))))

(defn slugify
  "Convert string to slug (URL-friendly format)."
  [input]
  (-> input
      str/lower-case
      (str/replace #"[^a-z0-9\s-]" "")
      (str/replace #"\s+" "-")
      (str/replace #"-+" "-")
      (str/replace #"^-|-$" "")))

(defn escape-xml
  "Escape XML special characters."
  [input]
  (escape-html input))

File: bindings/scala/src/main/scala/proven/SafeString.scala
// SPDX-License-Identifier: PMPL-1.0
// SPDX-FileCopyrightText: 2025 Hyperpolymath

package proven

import java.net.{URLEncoder, URLDecoder}
import java.nio.charset.StandardCharsets
import java.util.Base64
import scala.util.Try

/**
 * Safe string operations with XSS and injection prevention.
 */
object SafeString {

  private val HtmlEntities: Map[Char, String] = Map(
    '&' -> "&amp;",
    '<' -> "&lt;",
    '>' -> "&gt;",
    '"' -> "&quot;",
    '\'' -> "&#x27;"
  )

  /**
   * Escape string for safe HTML output (XSS prevention).
   */
  def escapeHtml(input: String): String = {
    val sb = new StringBuilder(input.length)
    input.foreach { c =>
      sb.append(HtmlEntities.getOrElse(c, c.toString))
    }
    sb.toString()
  }

  /**
   * Escape string for SQL single quotes.
   * Note: PREFER PARAMETERIZED QUERIES! This is for edge cases only.
   */
  def escapeSql(input: String): String = {
    input.replace("'", "''")
  }

  /**
   * Escape string for JavaScript string context.
   */
  def escapeJs(input: String): String = {
    val sb = new StringBuilder(input.length)
    input.foreach {
      case '\\' => sb.append("\\\\")
      case '\'' => sb.append("\\'")
      case '"' => sb.append("\\\"")
      case '\n' => sb.append("\\n")
      case '\r' => sb.append("\\r")
      case '\t' => sb.append("\\t")
      case '<' => sb.append("\\x3C")
      case '>' => sb.append("\\x3E")
      case c => sb.append(c)
    }
    sb.toString()
  }

  /**
   * Escape string for shell command (single-quote wrapping).
   */
  def escapeShell(input: String): String = {
    "'" + input.replace("'", "'\\''") + "'"
  }

  /**
   * Escape special characters for use in regex patterns.
   */
  def escapeRegex(input: String): String = {
    java.util.regex.Pattern.quote(input)
  }

  /**
   * URL-encode a string.
   */
  def urlEncode(input: String): String = {
    URLEncoder.encode(input, StandardCharsets.UTF_8.name())
  }

  /**
   * URL-decode a string.
   */
  def urlDecode(input: String): Option[String] = {
    Try(URLDecoder.decode(input, StandardCharsets.UTF_8.name())).toOption
  }

  /**
   * Base64-encode a string.
   */
  def base64Encode(input: String): String = {
    Base64.getEncoder.encodeToString(input.getBytes(StandardCharsets.UTF_8))
  }

  /**
   * Base64-decode a string.
   */
  def base64Decode(input: String): Option[String] = {
    Try {
      new String(Base64.getDecoder.decode(input), StandardCharsets.UTF_8)
    }.toOption
  }

  /**
   * Safely truncate a string to a maximum length.
   */
  def truncate(input: String, maxLength: Int, suffix: String = "..."): String = {
    if (maxLength <= 0) ""
    else if (input.length <= maxLength) input
    else if (maxLength <= suffix.length) input.take(maxLength)
    else input.take(maxLength - suffix.length) + suffix
  }

  /**
   * Strip HTML tags from string (basic - not a full parser).
   */
  def stripHtml(input: String): String = {
    input.replaceAll("<[^>]*>", "")
  }

  /**
   * Check if string contains only alphanumeric characters.
   */
  def isAlphanumeric(input: String): Boolean = {
    input.nonEmpty && input.forall(_.isLetterOrDigit)
  }

  /**
   * Check if string contains only ASCII characters.
   */
  def isAscii(input: String): Boolean = {
    input.forall(c => c >= 0 && c <= 127)
  }

  /**
   * Sanitize string to contain only specified allowed characters.
   */
  def sanitize(input: String, allowed: String = "a-zA-Z0-9_-"): String = {
    val pattern = s"[^$allowed]".r
    pattern.replaceAllIn(input, "")
  }

  /**
   * Remove control characters from string.
   */
  def removeControlChars(input: String): String = {
    input.filterNot(c =>
      (c >= 0 && c <= 8) || c == 11 || c == 12 || (c >= 14 && c <= 31) || c == 127
    )
  }

  /**
   * Normalize whitespace (collapse multiple spaces, trim).
   */
  def normalizeWhitespace(input: String): String = {
    input.replaceAll("\\s+", " ").trim
  }

  /**
   * Check if string looks like it contains injection attempts.
   */
  def containsSuspiciousPatterns(input: String): Boolean = {
    val lowerInput = input.toLowerCase
    val patterns = Seq(
      "<script",
      "javascript:",
      "union\\s+select",
      ";\\s*drop\\s+table"
    )

    patterns.exists { pattern =>
      pattern.r.findFirstIn(lowerInput).isDefined
    } || "on\\w+\\s*=".r.findFirstIn(lowerInput).isDefined
  }

  /**
   * Escape XML special characters.
   */
  def escapeXml(input: String): String = {
    escapeHtml(input) // Same entities for basic XML escaping
  }

  /**
   * Convert string to slug (URL-friendly format).
   */
  def slugify(input: String): String = {
    input
      .toLowerCase
      .replaceAll("[^a-z0-9\\s-]", "")
      .replaceAll("\\s+", "-")
      .replaceAll("-+", "-")
      .stripPrefix("-")
      .stripSuffix("-")
  }
}
